feature-libs/product-multi-dimensional/selector/components/guards/product-multi-dimensional-selector.guard.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { inject, Injectable } from '@angular/core';
+import { ActivatedRouteSnapshot, Router, UrlTree } from '@angular/router';
+import {
+  isNotUndefined,
+  Product,
+  ProductScope,
+  ProductService,
+  SemanticPathService,
+  VariantOption,
+} from '@spartacus/core';
+import { Observable, of } from 'rxjs';
+import { filter, map, switchMap, take } from 'rxjs/operators';
+
+/**
+ * Guard that ensures a product is valid and navigable within a multi-dimensional context.
+ *
+ * This service manages navigation and access control for products in a multi-dimensional product catalog.
+ * It verifies that the product specified in the route is valid and has the necessary attributes for display.
+ * If the product is not purchasable and has variant options, the guard attempts to find a valid product code
+ * from the available variants and redirects to the appropriate variant product URL.
+ *
+ * Without this guard, users could access a product detail page (PDP) for products that are not available for purchase.
+ */
+@Injectable({
+  providedIn: 'root',
+})
+export class ProductMultiDimensionalSelectorGuard {
+  protected productService: ProductService = inject(ProductService);
+  protected semanticPathService: SemanticPathService =
+    inject(SemanticPathService);
+  protected router: Router = inject(Router);
+
+  canActivate(
+    activatedRoute: ActivatedRouteSnapshot
+  ): Observable<boolean | UrlTree> {
+    const productCode = activatedRoute.params?.productCode;
+    if (!productCode) {
+      return of(false);
+    }
+
+    return this.productService
+      .get(productCode, ProductScope.MULTI_DIMENSIONAL_AVAILABILITY)
+      .pipe(
+        filter(isNotUndefined),
+        switchMap((multiDimensionalProduct: Product) => {
+          const isNotPurchasableAndHasVariantOptions =
+            !multiDimensionalProduct.purchasable &&
+            !!multiDimensionalProduct.variantOptions?.length;
+
+          return isNotPurchasableAndHasVariantOptions
+            ? this.findValidProductCodeAndReturnUrlTree(multiDimensionalProduct)
+            : of(multiDimensionalProduct.purchasable);
```

Comment:
```suggestion
          return isNotPurchasableAndHasVariantOptions
            ? this.findValidProductCodeAndReturnUrlTree(multiDimensionalProduct)
            : of(!!multiDimensionalProduct.purchasable);
```
To meet the return type of `canActivate` function

Reply:
fixed